Documentation for __remove_address_space in C++ for OpenCL (#647)
KhronosGroup/OpenCL-Docs: cxx4opencl/address_spaces.txt

gengenptr = reinterpret_cast<int**>(locgenptr); // legal.
gengenptr = reinterpret_cast<int**>(cnstgenptr); // legal.
----------

[[remove-addrspace]]
==== Address space removal

> Utility `+__remove address_space+` allows to derive types that have address

Reviewer: Since we haven't supported this in C++ for OpenCL 1.0 initially, I suggest we add this only in 2021 version? so maybe we can add some text to clarify that and then also dependency on https://github.com/KhronosGroup/OpenCL-Docs/pull/616?

> space qualifiers removed. The tool only throws away address space qualifiers,

Reviewer: ```suggestion
space qualifiers removed. The utility only removes an address space qualifier from a given type,
```

therefore, other type qualifiers such as const or volatile remain unchanged.

> [source,cpp]

Reviewer: I feel that the example alone is not enough, we should probably add at least a template class declaration:

`template<class T> struct remove_address_space;`

either following C++ spec style or reference pages:
https://www.cplusplus.com/reference/type_traits/remove_const/

Perhaps we should also point out that it is expected to have similar behavior to `remove_const` and others similar traits from C++17? Then it will be easier to explain and to understand the idea behind this feature.

> Utility `+__remove address_space+` allows to derive types that have address
> space qualifiers removed. The utility only removes an address space qualifier from a given type,
> therefore, other type qualifiers such as const or volatile remain unchanged.

Author: ```suggestion
C++ for OpenCL 2021 supports utility `+__remove address_space+` which allows to
derive types that have address space qualifiers removed. The utility only removes
an address space qualifier from a given type, therefore, other type qualifiers
such as `+const+` or `+volatile+` remain unchanged.
```

----------
template<typename T>

> void foo(T *par){

Reviewer 2: ```suggestion
void foo(T *par) {
```

  T var1; // error: function scope variable cannot be declared in global
          // address space
  __private T var2; // error: conflicting address space qualifiers are provided
                    // between types '__private T' and '__global int'
  __private __remove_address_space<T>::type var3; // type of var3 is __private int
}

> void bar(){

Reviewer 2: ```suggestion
void bar() {
```

>     __global int* ptr;

Reviewer 2: Indentation is inconsistent with template above (and in fact the rest of the document which seems to be using 2 spaces).

>     __global int* ptr;
>     foo(ptr);

Author: ```suggestion
  __global int* ptr;
  foo(ptr);
```

gengenptr = reinterpret_cast<int**>(locgenptr); // legal.
gengenptr = reinterpret_cast<int**>(cnstgenptr); // legal.
----------

[[remove-addrspace]]
==== Address space removal

> C++ for OpenCL 2021 supports utility `+__remove address_space+` which allows to
> derive types that have address space qualifiers removed. The utility only removes
> an address space qualifier from a given type, therefore, other type qualifiers
> such as `+const+` or `+volatile+` remain unchanged.

Author: ```suggestion
{cpp} for OpenCL 2021 supports utility `+__remove_address_space+` which allows to
derive types that have address space qualifiers removed. Its effect is analogous
to `+__remove_const+` and other similar traits in {cpp}17 `[meta.trans]`. The
utility only removes an address space qualifier from a given type, therefore, other
type qualifiers such as `+const+` or `+volatile+` remain unchanged.
```
